### PR TITLE
Functional coverage

### PR DESCRIPTION
**Pull request by christopherdaly-wf**
Added functional test coverage
@jayudey-wf

### CODE REVIEW

File: lib/src/task_process.dart [hunk 1]
```diff
@@ -58,6 +57,67 @@ class TaskProcess {
   Stream<String> get stderr => _stderr.stream;
   Stream<String> get stdout => _stdout.stream;
 
+  Future<List<int>> getChildPids(List<int> pids) async{
+    String executable = "pgrep";
+    var args = [];
+    for(int i in pids){
+      args.add("-P");
+      args.add("${i}");
```
Comment (jayudey-wf) [inline lib/src/task_process.dart]: i don't know if the curly braces are necessary here

File: lib/src/task_process.dart [hunk 2]
```diff
@@ -58,6 +57,67 @@ class TaskProcess {
...
+    }
+    TaskProcess pp = new TaskProcess(executable,args);
```
Comment (jayudey-wf) [inline lib/src/task_process.dart]: pp?

File: lib/src/task_process.dart [hunk 3]
```diff
@@ -58,6 +57,67 @@ class TaskProcess {
...
+    List<int> cpids = new List<int>();
+    pp.stdout.listen((l){
+      try{
+        int cpid = int.parse(l);
+        cpids.add(cpid);
+      }
+      on Exception{
```
Comment (jayudey-wf) [inline lib/src/task_process.dart]: just swallowing this?

File: lib/src/task_process.dart [hunk 4]
```diff
@@ -58,6 +57,67 @@ class TaskProcess {
...
+    var args = [];
+    for(int i in pids){
+      args.add("-P");
+      args.add("${i}");
+    }
+    TaskProcess pp = new TaskProcess(executable,args);
+    List<int> cpids = new List<int>();
+    pp.stdout.listen((l){
+      try{
+        int cpid = int.parse(l);
```
Comment (jayudey-wf) [inline lib/src/task_process.dart]: is this line necessary?

File: lib/src/task_process.dart [hunk 5]
```diff
@@ -58,6 +57,67 @@ class TaskProcess {
...
+      cpids = await getChildPids(cpids);
+      cpids.addAll(pids);
+      return cpids;
+    }
+    else return pids;
+  }
+
+  Future<bool> killAllChildren(int ppid) async{
+    List<int> cpids = await getChildPids([ppid]);
+    var args = ["-TERM"];
```
Comment (jayudey-wf) [inline lib/src/task_process.dart]: you could type this

File: lib/src/task_process.dart [hunk 6]
```diff
@@ -58,6 +57,67 @@ class TaskProcess {
...
+    cpids.remove(ppid);
+    for(int i in cpids)
+      args.add(i.toString());
+    TaskProcess k = new TaskProcess("kill",args);
+    await k.done;
+    if(await k.exitCode ==0)
```
Comment (jayudey-wf) [inline lib/src/task_process.dart]: spacing, you should run `pub run dart_dev format`

File: lib/src/task_process.dart [hunk 7]
```diff
@@ -58,6 +57,67 @@ class TaskProcess {
...
+      return true;
+    else
+      return false;
+
```
Comment (jayudey-wf) [inline lib/src/task_process.dart]: weird space

File: lib/src/task_process.dart [hunk 8]
```diff
@@ -58,6 +57,67 @@ class TaskProcess {
...
+//      });
+//      a.stderr.listen((l) {
+//        print(l);
+//      });
+//
+//    await a.done;
+//    try{_process.kill();}
+//          on Exception{
+//
+//          }
```
Comment (jayudey-wf) [inline lib/src/task_process.dart]: should this be removed?

File: lib/src/tasks/coverage/api.dart [hunk 9]
```diff
@@ -128,39 +135,44 @@ class CoverageTask extends Task {
...
 
+  TaskProcess _seleniumServerProcess;
+
   /// Directory to output all coverage related artifacts.
   Directory _outputDirectory;
 
   /// List of directories on which coverage should be reported.
   List<String> _reportOn;
 
+  void _testFileValidation(path,fileCollection) {
```
Comment (jayudey-wf) [inline lib/src/tasks/coverage/api.dart]: these params could be typed

File: lib/src/tasks/coverage/api.dart [hunk 10]
```diff
@@ -233,11 +246,88 @@ class CoverageTask extends Task {
...
+        continue;
+      }
+
+      List<int> validPorts = new List<int>();
+      List<Future> wsDone = new List<Future>();
+
+      //Check for observatories with actual data
+      for(int port in observatoryPorts) {
+        try {
+          var ws = await WebSocket.connect("ws://127.0.0.1:${port}/ws");
```
Comment (jayudey-wf) [inline lib/src/tasks/coverage/api.dart]: type?

File: lib/src/tasks/coverage/cli.dart [hunk 11]
```diff
@@ -54,8 +57,9 @@ class CoverageCli extends TaskCli {
 
     bool unit = parsedArgs['unit'];
     bool integration = parsedArgs['integration'];
+    bool functional = parsedArgs['functional'];
 
-    if (!unit && !integration) {
+    if (!unit && !integration && !functional) {
       return new CliResult.fail(
           'No tests were selected. Include at least one of --unit or --integration.');
```
Comment (jayudey-wf) [inline lib/src/tasks/coverage/cli.dart]: add functional flag

File: test/integration/coverage_test.dart [hunk 12]
```diff
@@ -67,6 +70,12 @@ void main() {
       expect(lcov.existsSync(), isTrue);
     }, timeout: new Timeout(new Duration(seconds: 60)));
 
+    test('should generate coverage for Functional tests', () async{
+      expect(await runCoverage(projectWithFunctionalTests,functional: true), isTrue);
+      File lcov = new File('$projectWithBrowserTests/coverage/coverage.lcov');
```
Comment (jayudey-wf) [inline test/integration/coverage_test.dart]: i think this should be projectWithFunctionalTests

File: test/fixtures/coverage/functional_test/functional/pubspec.yaml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+name: functional_test
+version: 0.0.0
+dev_dependencies:
+  unittest: '>=0.11.5+4 < 1.0.0'
+  webdriver: '0.9.1'
```
Comment (jayudey-wf) [inline test/fixtures/coverage/functional_test/functional/pubspec.yaml]: you could add last lines to a couple of these files

File: test/fixtures/coverage/functional_test/functional/test/simple_test.dart [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+void main() {
+  WebDriver driver;
+
+  setUp(() async{
+    driver = await WebDriver.createDriver(
+        desiredCapabilities: {'browserName': 'chrome','chromeOptions':{'binary': '/usr/local/bin/dartium'}});
+  });
+
+//  tearDown(() => driver.quit());
```
Comment (jayudey-wf) [inline test/fixtures/coverage/functional_test/functional/test/simple_test.dart]: i would leave a comment in here about why you aren't doing this too

File: lib/src/tasks/coverage/api.dart [hunk 15]
```diff
@@ -26,6 +26,9 @@ import 'package:dart_dev/src/tasks/coverage/config.dart';
 import 'package:dart_dev/src/tasks/coverage/exceptions.dart';
 import 'package:dart_dev/src/tasks/task.dart';
 
+
+import 'package:dart_dev/src/tasks/config.dart';
```
Comment (jayudey-wf) [inline lib/src/tasks/coverage/api.dart]: you could move this import and alphabetize it

File: lib/src/tasks/coverage/api.dart [hunk 16]
```diff
@@ -128,39 +135,44 @@ class CoverageTask extends Task {
   /// the coverage collection has completed.
   TaskProcess _lastTestProcess;
 
+  TaskProcess _seleniumServerProcess;
+
   /// Directory to output all coverage related artifacts.
   Directory _outputDirectory;
 
   /// List of directories on which coverage should be reported.
   List<String> _reportOn;
```
Comment (jayudey-wf) [inline lib/src/tasks/coverage/api.dart]: this method should be moved below and sorted with the other local method

File: lib/src/tasks/coverage/api.dart [hunk 17]
```diff
@@ -233,11 +246,88 @@ class CoverageTask extends Task {
...
+
+      List<int> validPorts = new List<int>();
+      List<Future> wsDone = new List<Future>();
+
+      //Check for observatories with actual data
+      for(int port in observatoryPorts) {
+        try {
+          var ws = await WebSocket.connect("ws://127.0.0.1:${port}/ws");
+          wsDone.add(ws.done);
+          ws.add("{\"id\":\"3\",\"method\":\"getVM\",\"params\":{}}");
```
Comment (jayudey-wf) [inline lib/src/tasks/coverage/api.dart]: will this always be 3?

File: lib/src/tasks/coverage/api.dart [hunk 18]
```diff
@@ -233,11 +246,88 @@ class CoverageTask extends Task {
...
+        _killTest();
+        collections.add(collection);
+//        if (await _lastTestProcess.exitCode > 0) continue;
+      }
+      await _seleniumServerProcess.killGroup();
+    }
+
+    return collections;
+    // Merge all individual coverage collection files into one.
+//    _collection = _merge(collections);
```
Comment (jayudey-wf) [inline lib/src/tasks/coverage/api.dart]: inconsistent commenting spacing

File: lib/src/tasks/coverage/api.dart [hunk 19]
```diff
@@ -317,13 +407,79 @@ class CoverageTask extends Task {
...
     }
 
-    await _collect();
+    TaskProcess serve;
+
+    if(functional_tests.isNotEmpty){
+      //    String executable = 'selenium-server';
+//    int port = await getOpenPort();
+      List args = ["-port"
+//    ,"${port}"
```
Comment (jayudey-wf) [inline lib/src/tasks/coverage/api.dart]: remove un-used items?

File: lib/src/tasks/coverage/api.dart [hunk 20]
```diff
@@ -317,13 +407,79 @@ class CoverageTask extends Task {
...
+          serverRunning.complete();
+        }
+      });
+
+      _seleniumServerProcess = new TaskProcess('selenium-server', args);
+
+      RegExp _observatoryPortPattern = new RegExp(
+          r'Observatory listening (at|on) http:\/\/127\.0\.0\.1:(\d+)');
+
+      int portPatternCount=0;
```
Comment (jayudey-wf) [inline lib/src/tasks/coverage/api.dart]: un-used variable?

**Comment by jayudey-wf** [comment]
first pass complete

File: lib/src/tasks/coverage/cli.dart [hunk 21]
```diff
@@ -64,12 +68,17 @@ class CoverageCli extends TaskCli {
     bool open = TaskCli.valueOf('open', parsedArgs, true);
 
     List<String> tests = [];
+    List<String> functional_tests = [];
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/cli.dart]: Variables should be camelCased: `functionalTests`

File: lib/src/tasks/coverage/cli.dart [hunk 22]
```diff
@@ -80,18 +89,27 @@ class CoverageCli extends TaskCli {
             'This project does not specify any integration tests.');
       }
     }
+    if(functional_tests.isEmpty){
+      if(functional && config.test.functionalTests.isEmpty){
+        return new CliResult.fail(
+            'This project does not specify any functional tests.');
+      }
+    }
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/cli.dart]: This if block should be nested within the above `if (tests.isEmpty) {` block because we only want to fail when no tests were specified, whereas this will fail when no functional tests are specified, even if other tests (unit, integration) were.

File: lib/src/tasks/coverage/config.dart [hunk 23]
```diff
@@ -17,6 +17,7 @@ library dart_dev.src.tasks.coverage.config;
 import 'package:dart_dev/src/tasks/config.dart';
 
 const bool defaultHtml = true;
+const List<String> defaultFunctionalTests = const [];
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/config.dart]: This option should be a part of the test task configuration, not coverage. The coverage task uses the test configuration as well as its own

File: lib/src/tasks/test/cli.dart [hunk 24]
```diff
@@ -92,8 +96,7 @@ class TestCli extends TaskCli {
     if (hasRestParams(parsedArgs)) {
       individualTests = await addToTestsFromRest(tests, parsedArgs.rest);
     }
-
-    if (isExplicitlyFalse(unit) && !integration && individualTests == 0) {
+    if (isExplicitlyFalse(unit) && !integration && !functional && individualTests == 0) {
       return new CliResult.fail(
           'No tests were selected. Include at least one of --unit or --integration.');
```
Comment (evanweible-wf) [inline lib/src/tasks/test/cli.dart]: Update this message to include the `--functional` flag.

File: lib/src/tasks/coverage/config.dart (hunk 23 above)
Comment (evanweible-wf) [inline lib/src/tasks/coverage/config.dart]: Looks like you do actually have this in the test task configuration, so you should be able to use that one instead of this

File: test/fixtures/coverage/functional_test/tool/dev.dart [hunk 25]
```diff
@@ -0,0 +1,9 @@
+library tool.dev;
+
+import 'package:dart_dev/dart_dev.dart' show dev, config;
+
+main(List<String> args) async {
+  config.test.functionalTests = ['functional'];
```
Comment (evanweible-wf) [inline test/fixtures/coverage/functional_test/tool/dev.dart]: All tests should be in the `test/` directory, so instead of a directory structure like `project/functional/test/`, it should be `project/test/functional`. This fits with the other test suites as well:

```
project/test/unit/
project/test/integration/
project/test/functional/
```

File: lib/src/tasks/coverage/api.dart [hunk 26]
```diff
@@ -475,4 +631,45 @@ class CoverageTask extends Task {
       return port;
     }
   }
+
+  Future<List<int>> _test_functional(File file) async {
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: Is `file` here an HTML file or a Dart file? Or do both work?

File: lib/src/tasks/coverage/api.dart [hunk 27]
```diff
@@ -475,4 +631,45 @@ class CoverageTask extends Task {
...
+    int count = observatoryPort.length;
+
+    String executable = 'pub';
+    List args = ['run',
+//        int port = await getOpenPort();
+//        String executable = 'dart';
+//        List args = ["--observe=${port}"
+    file.path.split(config.test.functionalTests[0] + '/')[1]];
+    print(config.test.functionalTests[0]);
+    TaskProcess process = new TaskProcess(executable, args,workingDirectory:config.test.functionalTests[0]);
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: Is the only difference between running a functional test and a regular test that the functional test is run as a child process via `pub run`? Could every test be run like this?

File: lib/src/tasks/coverage/api.dart [hunk 28]
```diff
@@ -66,11 +70,12 @@ class CoverageTask extends Task {
   /// If [html] is true, `genhtml` will be used to generate an HTML report of
   /// the collected coverage and the report will be opened.
   static Future<CoverageResult> run(List<String> tests,
-      {bool html: defaultHtml,
+      {List<String> functional_tests:defaultFunctionalTests,
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: The idea behind the design of the coverage task right now is that it doesn't care what type of test is being run (unit, integration, or functional) - it just receives a list of tests to run and collect coverage on. This breaks that design by making functional tests a special case. I don't know enough about how functional tests are run yet, so it may be the case that this is necessary. But, I'd like to avoid this if possible. I'll leave some comments/questions about new methods below specific to functional tests

File: lib/src/tasks/coverage/api.dart [hunk 29]
```diff
@@ -317,13 +407,79 @@ class CoverageTask extends Task {
...
+        }
+      });
+      serve.stdout.listen((line){
+        print(line);
+        if(_serving.hasMatch(line)){
+          serverRunning.complete();
+        }
+      });
+
+      _seleniumServerProcess = new TaskProcess('selenium-server', args);
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: If possible, I'd like to avoid adding this dependency to this task. This forces the coverage task to be aware of functional tests and their requirement on a selenium server. While that is the case for our usage, it's possible that other consumers may run functional tests in a different way and without the need for a selenium server. Could we instead make this the responsibility of the project via the `before`/`after` configuration for the coverage task? For example..

```dart
main() async {
  config.coverage
    ..before = [_startSeleniumServer]
    ..after = [_stopSeleniumServer];
}

_startSeleniumServer() async { ... }
_stopSeleniumServer() async { ... }
```

File: lib/src/tasks/coverage/api.dart [hunk 30]
```diff
@@ -317,13 +407,79 @@ class CoverageTask extends Task {
...
+    if(functional_tests.isNotEmpty){
+      //    String executable = 'selenium-server';
+//    int port = await getOpenPort();
+      List args = ["-port"
+//    ,"${port}"
+      ];
+      args = [];
+      observatoryPort = new List<int>();
+      RegExp _serving = new RegExp("Serving .* on http:\/\/localhost:8080");
+      serve = new TaskProcess('pub', ['serve']);
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: Similar note here.. I'd rather this not be something that's different for functional tests. I think there could be some convergence here with @jayudey-wf's work to run all tests during the coverage task via a pub serve process. If all tests were run like this, then we're another step closer to keeping this coverage task "dumber" by not needing to run functional tests differently than other tests

File: lib/src/tasks/coverage/api.dart [hunk 31]
```diff
@@ -475,4 +631,45 @@ class CoverageTask extends Task {
...
+
+    String _testsFailedPattern = 'Some tests failed.';
+    RegExp _testsPassedPat = new RegExp(r'All\s[\d]*\s?tests passed.');
+
+    process.stderr.listen((l) => _coverageErrorOutput.add('    $l'));
+
+    await for (String line in process.stdout) {
+      _coverageOutput.add('    $line');
+      if (line.contains(_observatoryFailPattern)) {
+        throw new CoverageTestSuiteException(file.path);
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: Most of the logic in this area is duplicated from how all other tests are run.. I think if we can move the dependency on selenium out of this task and run all tests via a pub serve process, then we could eliminate the need for this separate test execution & coverage collection path for functional tests

File: lib/src/tasks/coverage/api.dart (hunk 30 above)
Comment (jayudey-wf) [inline lib/src/tasks/coverage/api.dart]: i think for now this is necessary because currently functional tests need to be run in a separate directory with their own pubspec.yaml (when webdriver.dart and w_webdriver_utils get off of the unittest package I think this might work)

Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: Any idea how far away that work is?

File: test/fixtures/coverage/functional_test/tool/dev.dart (hunk 25 above)
Comment (evanweible-wf) [inline test/fixtures/coverage/functional_test/tool/dev.dart]: I guess this is because of the requirement that functional tests currently must be run with the old test package... bummer

**Comment by rmconsole2-wf** [comment]
@christopherdaly-wf This pull request has merge conflicts, please resolve.

**Comment by rmconsole3-wf** [comment]
@christopherdaly-wf This pull request has merge conflicts, please resolve.

**Comment by rmconsole2-wf** [comment]
@christopherdaly-wf This pull request has merge conflicts, please resolve.

**Comment by christopherdaly-wf** [comment]
@evanweible-wf Could you give this another review?

File: README.md [hunk 32]
```diff
@@ -130,6 +130,23 @@ bashcompinit
 source <path/to/ddev-completion.sh>
 ```
 
+#### Functional Test Code Coverage
+
+>The following dependencies are needed to run functional code coverage:
+>
+>Chromedriver v2.14
```
Comment (jayudey-wf) [inline README.md]: i thought i had to go much further back in order to get this to work?

File: README.md [hunk 33]
```diff
@@ -130,6 +130,23 @@ bashcompinit
...
+>https://chromedriver.storage.googleapis.com/index.html?path=2.14/
+>
+>Selenium Server
+>http://www.seleniumhq.org/download/
+
+In order for coverage to be collected against functional tests, Dartium must be run against native dart code.  To launch an instance of Dartium with chromedriver you will need to configure the chromedriver instance to include the path to your dartium binary.
+```
+WebDriver driver = await createDriver(
+	desired: {'browserName': 'chrome','chromeOptions':{'binary': dartpath}});
+```
```
Comment (jayudey-wf) [inline README.md]: you should add a note that it can be different based on how you are instantiating them (using w_webdriver_utils or just webdriver.dart)

File: README.md [hunk 34]
```diff
@@ -130,6 +130,23 @@ bashcompinit
...
+>Chromedriver v2.14
+>https://chromedriver.storage.googleapis.com/index.html?path=2.14/
+>
+>Selenium Server
+>http://www.seleniumhq.org/download/
+
+In order for coverage to be collected against functional tests, Dartium must be run against native dart code.  To launch an instance of Dartium with chromedriver you will need to configure the chromedriver instance to include the path to your dartium binary.
+```
+WebDriver driver = await createDriver(
+	desired: {'browserName': 'chrome','chromeOptions':{'binary': dartpath}});
```
Comment (jayudey-wf) [inline README.md]: dartiumpath*

File: README.md [hunk 35]
```diff
@@ -130,6 +130,23 @@ bashcompinit
...
+```
+The `coverage` and `test` tasks will both handle closing your browser instances after functional test completion.  In fact, this is necessary for collection of coverage data.  Because of this, you will need to omit the standard `tearDown(() => driver.quit());` from your test instances.
```
Comment (jayudey-wf) [inline README.md]: i would strongly emphasize the fact that you need to update your tests to not close

File: README.md [hunk 36]
```diff
@@ -130,6 +130,23 @@ bashcompinit
...
+
+>The following dependencies are needed to run functional code coverage:
+>
+>Chromedriver v2.14
+>https://chromedriver.storage.googleapis.com/index.html?path=2.14/
+>
+>Selenium Server
+>http://www.seleniumhq.org/download/
+
+In order for coverage to be collected against functional tests, Dartium must be run against native dart code.  To launch an instance of Dartium with chromedriver you will need to configure the chromedriver instance to include the path to your dartium binary.
```
Comment (jayudey-wf) [inline README.md]: I would maybe re-word this to be something like to the effect of. 'the application under test must be utilizing native dart code'

File: lib/src/tasks/coverage/api.dart [hunk 37]
```diff
@@ -537,8 +550,32 @@ class CoverageTask extends Task {
...
+      if (server != null && server.observatoryPort.length > 0) {
+        List<int> result = await server.checkPorts();
+        return result;
+      }
+      return [port];
+    }
+  }
 
-      return port;
+  void _testFileValidation(String path, List<File> fileCollection) {
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: This is a bit of an anti-pattern - I would expect this method to return a list of files instead of populating a list that gets passed in.

File: setup_dartium.sh [hunk 38]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
```
Comment (evanweible-wf) [inline setup_dartium.sh]: Could we move this to the `tool/` directory?

File: setup_dartium.sh [hunk 39]
```diff
@@ -0,0 +1,44 @@
...
+set -e
+
+DART_DIST="dartsdk-linux-x64-release.zip";
+DARTIUM_DIST="dartium-linux-x64-release.zip";
+SELENIUM_JAR="selenium-server.jar";
+
+echo "Fetching dart sdk and Dartium
"
+
+curl "http://storage.googleapis.com/dart-archive/channels/stable/release/latest/sdk/$DART_DIST" > $DART_DIST
```
Comment (evanweible-wf) [inline setup_dartium.sh]: Why does the sdk need to be downloaded again? Isn't it already available on travis? I think I understand dartium since it's used in a symlink below

File: lib/src/task_process.dart [hunk 40]
```diff
@@ -58,6 +57,43 @@ class TaskProcess {
   Stream<String> get stderr => _stderr.stream;
   Stream<String> get stdout => _stdout.stream;
 
+  Future<List<int>> getChildPids(List<int> pids) async {
+    String executable = "pgrep";
```
Comment (evanweible-wf) [inline lib/src/task_process.dart]: Single quotes

File: README.md [hunk 41]
```diff
@@ -130,6 +130,41 @@ bashcompinit
 source <path/to/ddev-completion.sh>
 ```
 
+#### Functional Test Code Coverage
+
+The following dependencies are needed to run functional code coverage:
+
+Chromedriver 2.14:
+```
+https://chromedriver.storage.googleapis.com/index.html?path=2.14/
```
Comment (evanweible-wf) [inline README.md]: Is 2.14 a strict requirement? Or would the latest version work (2.20)? Just wondering because there's a brew formula to install this that would be easier than doing it manually:

```bash
brew install chromedriver
```

File: lib/src/tasks/coverage/exceptions.dart [hunk 42]
```diff
@@ -38,3 +38,10 @@ class MissingLcovException implements Exception {
   final String message = missingLcovMessage;
   String toString() => 'MissingLcovException: $message';
 }
+
+class PortBoundException implements Exception {
+  final String message;
+  PortBoundException(String message)
+      : this.message = 'Port could not be bound: $message';
+  String toString() => 'PortBoundException: $message';
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/exceptions.dart]: This exception should include the port number. When the selenium server fails to start (because it's already running), the port number it tried to connect to should be output with the error to make it easy to kill the process.

File: .travis.yml [hunk 43]
```diff
@@ -1,9 +1,12 @@
 language: dart
+#sudo: required
```
Comment (evanweible-wf) [inline .travis.yml]: Remove this if it's not needed

File: test/integration/examples_test.dart [hunk 44]
```diff
@@ -33,9 +33,10 @@ Future<bool> serveExamplesFor(String projectPath) async {
   bool served = false;
 
   Pattern pubServePattern = new RegExp(r'pub serve .* example');
-  process.stdout.listen((line) {
+  process.stdout.listen((line) async {
     if (line.contains(pubServePattern)) {
       served = true;
+      await process.killGroup();
```
Comment (evanweible-wf) [inline test/integration/examples_test.dart]: Why was this added?

File: README.md [hunk 45]
```diff
@@ -335,6 +337,18 @@ configuration from the `config.test` object.
...
+            <td><code>seleniumCommand</code></td>
+            <td><code>String</code></td>
+            <td><code>"selenium-server"</code></td>
+            <td>Command used to execute Selenium for functional testing.</td>
+        </tr>
+        <tr>
+            <td><code>seleniumSuccess</code></td>
+            <td><code>String</code></td>
+            <td><code>"Selenium Server is up and running"</code></td>
+            <td>Value used to verify that Selenium has been started successfully.  This value is accurate for v2.48.0 of Selenium Server.</td>
```
Comment (evanweible-wf) [inline README.md]: Oops, these config options are no longer included. I should have removed this

File: test/integration/coverage_test.dart [hunk 46]
```diff
@@ -37,10 +40,13 @@ Future<bool> runCoverage(String projectPath, {bool html: false}) async {
   }
 
   List args = ['run', 'dart_dev', 'coverage'];
+  if (functional) {
+    args.add('--functional');
+    args.add('--no-unit');
```
Comment (evanweible-wf) [inline test/integration/coverage_test.dart]: Can remove this `--no-unit` line now that I've updated the text fixture's dart_dev config to set its unit tests suite to an empty list.

File: test/integration/coverage_test.dart [hunk 47]
```diff
@@ -78,6 +84,14 @@ void main() {
       expect(lcov.existsSync(), isTrue);
     }, timeout: new Timeout(new Duration(seconds: 60)));
 
+    test('should generate coverage for Functional tests', () async {
+      expect(await runCoverage(projectWithFunctionalTests, functional: true),
+          isTrue);
+      File lcov =
+          new File('$projectWithFunctionalTests/coverage/coverage.lcov');
+      expect(lcov.existsSync(), isTrue);
+    }, timeout: new Timeout(new Duration(seconds: 300)));
```
Comment (evanweible-wf) [inline test/integration/coverage_test.dart]: I imagine this timeout was for testing purposes? Probably should be removed

File: test/integration/examples_test.dart [hunk 48]
```diff
@@ -33,9 +33,10 @@ Future<bool> serveExamplesFor(String projectPath) async {
   bool served = false;
 
   Pattern pubServePattern = new RegExp(r'pub serve .* example');
-  process.stdout.listen((line) {
+  process.stdout.listen((line) async {
     if (line.contains(pubServePattern)) {
       served = true;
+      await process.killAllChildren();
```
Comment (evanweible-wf) [inline test/integration/examples_test.dart]: Remove this and remove the `async` keyword above

File: .travis.yml (hunk 43 above)
Comment (evanweible-wf) [inline .travis.yml]: Still need to remove this

**Comment by evanweible-wf** [comment]
@christopherdaly-wf just a few things to clean up and then I think this will be ready to go

**Comment by rmconsole2-wf** [comment]
@christopherdaly-wf This pull request has merge conflicts, please resolve.

**Comment by evanweible-wf** [comment]
@christopherdaly-wf #126 has been merged, so a merge with latest master should fix the CI failure

**Comment by rmconsole3-wf** [comment]
@christopherdaly-wf This pull request has merge conflicts, please resolve.

**Comment by rmconsole3-wf** [comment]
@christopherdaly-wf This pull request has merge conflicts, please resolve.

**Comment by rmconsole2-wf** [comment]
@christopherdaly-wf This pull request has merge conflicts, please resolve.
